### ISSUE TITLE
Temp View Generation does not properly work with Spark Connect

### ISSUE THREAD

**Issue by grundprinzip** [pyspark-ai/pyspark-ai#75]
Since Spark Connect will lazily evaluate the generated code, using the same name for the view for every invocaiton of `ai.transform()` will not work.

The current behavior works because the execution plan is eagerly evaluated when `spark.sql()` is called, but this will not work properly, when the analysis of the plan is deferred.

Repro

```
> env PYSPARK_DRIVER_PYTHON=ipython poetry run pyspark --remote local --packages org.apache.spark:spark-connect_2.12:3.4.1
```

```
In [7]: from pyspark_ai import SparkAI
   ...: from langchain.chat_models import ChatOpenAI
   ...:
   ...: llm = ChatOpenAI(model="gpt-3.5-turbo")
   ...: ai = SparkAI(llm=llm, spark_session=spark)
   ...: ai.activate()
   ...:
   ...: df = spark.range(10)
   ...: df2 = spark.range(100)
   ...:
   ...: r = df.ai.transform("count of rows")
   ...: assert(r.collect()[0][0] == 10)
   ...:
   ...: r2 = df2.ai.transform("count of rows")
   ...:
   ...: # Attention
   ...: assert(r2.collect()[0][0] == 100)
   ...: assert(r.collect()[0][0] == 10)
INFO: Creating temp view for the transform:
df.createOrReplaceTempView("temp_view_for_transform")

2023-07-23 11:40:45,277 48026 INFO execute_command Execute command for command create_dataframe_view { input { common { plan_id: 20 } range { start: 0 end: 10 step: 1 } } name: "temp_view_for_transform" replace: true }
2023-07-23 11:40:45,277 48026 INFO _execute_and_fetch ExecuteAndFetch
2023-07-23 11:40:45,277 48026 INFO _execute_and_fetch_as_iterator ExecuteAndFetchAsIterator
2023-07-23 11:40:45,287 48026 INFO schema Schema for plan: root { common { plan_id: 20 } range { start: 0 end: 10 step: 1 } }
INFO: SQL query for the transform:
SELECT COUNT(*) FROM temp_view_for_transform

2023-07-23 11:40:46,064 48026 INFO execute_command Execute command for command sql_command { sql: "SELECT COUNT(*) FROM temp_view_for_transform" }
2023-07-23 11:40:46,064 48026 INFO _execute_and_fetch ExecuteAndFetch
2023-07-23 11:40:46,064 48026 INFO _execute_and_fetch_as_iterator ExecuteAndFetchAsIterator
2023-07-23 11:40:46,074 48026 DEBUG _execute_and_fetch_as_iterator Received the SQL command result.
2023-07-23 11:40:46,079 48026 DEBUG _execute_and_fetch_as_iterator Received metric batch.
2023-07-23 11:40:46,080 48026 INFO to_table Executing plan root { common { plan_id: 24 } sql { query: "SELECT COUNT(*) FROM temp_view_for_transform" } }
2023-07-23 11:40:46,080 48026 INFO _execute_and_fetch ExecuteAndFetch
2023-07-23 11:40:46,080 48026 INFO _execute_and_fetch_as_iterator ExecuteAndFetchAsIterator
2023-07-23 11:40:46,084 48026 DEBUG _execute_and_fetch_as_iterator Received the schema.
2023-07-23 11:40:46,108 48026 DEBUG _execute_and_fetch_as_iterator Received arrow batch rows=1 size=328
2023-07-23 11:40:46,109 48026 DEBUG _execute_and_fetch_as_iterator Received metric batch.
INFO: Creating temp view for the transform:
df.createOrReplaceTempView("temp_view_for_transform")

2023-07-23 11:40:46,109 48026 INFO execute_command Execute command for command create_dataframe_view { input { common { plan_id: 21 } range { start: 0 end: 100 step: 1 } } name: "temp_view_for_transform" replace: true }
2023-07-23 11:40:46,110 48026 INFO _execute_and_fetch ExecuteAndFetch
2023-07-23 11:40:46,110 48026 INFO _execute_and_fetch_as_iterator ExecuteAndFetchAsIterator
2023-07-23 11:40:46,116 48026 INFO schema Schema for plan: root { common { plan_id: 21 } range { start: 0 end: 100 step: 1 } }
INFO: SQL query for the transform:
SELECT COUNT(*) FROM temp_view_for_transform

2023-07-23 11:40:46,119 48026 INFO execute_command Execute command for command sql_command { sql: "SELECT COUNT(*) FROM temp_view_for_transform" }
2023-07-23 11:40:46,119 48026 INFO _execute_and_fetch ExecuteAndFetch
2023-07-23 11:40:46,119 48026 INFO _execute_and_fetch_as_iterator ExecuteAndFetchAsIterator
2023-07-23 11:40:46,123 48026 DEBUG _execute_and_fetch_as_iterator Received the SQL command result.
2023-07-23 11:40:46,128 48026 DEBUG _execute_and_fetch_as_iterator Received metric batch.
2023-07-23 11:40:46,128 48026 INFO to_table Executing plan root { common { plan_id: 27 } sql { query: "SELECT COUNT(*) FROM temp_view_for_transform" } }
2023-07-23 11:40:46,128 48026 INFO _execute_and_fetch ExecuteAndFetch
2023-07-23 11:40:46,128 48026 INFO _execute_and_fetch_as_iterator ExecuteAndFetchAsIterator
2023-07-23 11:40:46,132 48026 DEBUG _execute_and_fetch_as_iterator Received the schema.
2023-07-23 11:40:46,151 48026 DEBUG _execute_and_fetch_as_iterator Received arrow batch rows=1 size=328
2023-07-23 11:40:46,152 48026 DEBUG _execute_and_fetch_as_iterator Received metric batch.
2023-07-23 11:40:46,152 48026 INFO to_table Executing plan root { common { plan_id: 24 } sql { query: "SELECT COUNT(*) FROM temp_view_for_transform" } }
2023-07-23 11:40:46,152 48026 INFO _execute_and_fetch ExecuteAndFetch
2023-07-23 11:40:46,152 48026 INFO _execute_and_fetch_as_iterator ExecuteAndFetchAsIterator
2023-07-23 11:40:46,156 48026 DEBUG _execute_and_fetch_as_iterator Received the schema.
2023-07-23 11:40:46,174 48026 DEBUG _execute_and_fetch_as_iterator Received arrow batch rows=1 size=328
2023-07-23 11:40:46,174 48026 DEBUG _execute_and_fetch_as_iterator Received metric batch.
---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
Cell In[7], line 18
     16 # Attention
     17 assert(r2.collect()[0][0] == 100)
---> 18 assert(r.collect()[0][0] == 10)

AssertionError:

In [8]:
```

**Comment by gengliangwang** [comment]
@grundprinzip the issue sounds more like from Spark connect? For example, given `df` and `df2`:
```
df.createOrReplaceTempView("temp_view_for_transform")
df2.createOrReplaceTempView("temp_view_for_transform")
```
The temp view should be based on `df2` instead of `df1`, right?

**Comment by grundprinzip** [comment]
Yes, that's what I meant when I said in regular PySpark the view is eagerly evaluated and the plan is created, whereas in Spark Connect the view is created, but we don't immediately resolve it to a plan.


Let me give you another example: Imagine you have two threads running the transform in parallel, you will have cases where the view resolution will pick up the wrong temp view depending on ordering.

The reason why it works fine serially is because when you run `spark.sql` it immediately plans the query and drops the sql statement and attaches a final execution plan with the dataframe. In Spark Connect that is not the case, we late-bind the SQL statement and wait for the action to execute it.

**Comment by gengliangwang** [comment]
@grundprinzip Yes I was wondering why Spark Connect has to be different. (Probably it is by design to make the remote function calls lightweight. )

I guess the simplest way to fix it is to generate random temp view names.

**Comment by gengliangwang** [comment]
@grundprinzip https://github.com/databrickslabs/pyspark-ai/pull/80 should fix this one, right?
